contracts/test/helpers.js

```diff
@@ -752,6 +753,48 @@ async function differenceInStrategyBalance(
   return returnVals;
 }
 
+async function addActualBalancesToSquidData(squidDataCsvFile, outputFileName, tokenContract) {
+  // CSV file in format account, balance, blockNumber (balance from squid)
+  const data = fs.readFileSync(squidDataCsvFile, "utf8").split('\n');
+  const combinedData = [];
+  //const blockNumber = await tokenContract.provider.getBlockNumber();
+  for (let i = 0; i < data.length; i++) {
+    const [account, balance,] = data[i].split(',');
+    const expectedBalance = BigNumber.from(balance);
+    const actualBalance = await tokenContract.balanceOf(account)
+
+    combinedData.push([account, balance, actualBalance.toString(10)].join(','));
+    if (i % 50 == 0) {
+      console.log(`Inspected ${i + 1} accounts`);
+    }
+    if (!actualBalance.eq(expectedBalance)) {
+      //console.log(`Balance miss match ${account} actual: ${actualBalance.toString(10)} expected: ${expectedBalance.toString(10)}`)
+    }
+  }
+  const csvContent = combinedData.join('\n');
+  fs.writeFileSync(outputFileName, csvContent);
+}
+
+async function compareUpgradedContractBalances(balancesFile, upgradedTokenContract) {
+  // CSV file in format account, squidBalance, onChainBalance
+  const data = fs.readFileSync(balancesFile, "utf8").split('\n');
+  let balanceMissmatches = 0;
+  for (let i = 0; i < data.length; i++) {
+    const [account,,balanceBefore] = data[i].split(',');
+    const expectedBalance = BigNumber.from(balanceBefore);
+    const actualBalance = await upgradedTokenContract.balanceOf(account)
```

Comment:
Would be better to do this in batches using `Promise.all`

Reply:
Agreed, I didn't do that in the end because all the storage slots are already cached and this function executes quite quickly